### PR TITLE
Arm64 multicore support - simplified code generation

### PR DESCRIPTION
**Pull request by xavierleroy**
See discussion at https://github.com/ocaml/ocaml/pull/10972

### CODE REVIEW

File: asmcomp/arm64/emit.mlp [hunk 1]
```diff
@@ -468,21 +468,18 @@ module BR = Branch_relaxation.Make (struct
...
-    | Lop (Iload  { memory_chunk; addressing_mode; is_atomic }) ->
-      let based = match addressing_mode with Iindexed _ -> 0 | Ibased _ -> 1 in
-      based + begin match memory_chunk with Single -> 2 | _ -> 1 end
-            + (if is_atomic then 1 else 0)
-    | Lop (Istore ( memory_chunk, addressing_mode, initialization )) ->
-      let word_non_init_add =
-        match memory_chunk, addressing_mode, initialization with
-          Word_val, Iindexed ofs, false when ofs != 0 -> 1
-        | Word_val, Ibased (_, ofs), false when ofs != 0 -> 1
+    | Lop (Iload  { memory_chunk; addressing_mode }) ->
```
Comment (ctk21) [inline asmcomp/arm64/emit.mlp]: It it correct that we can ignore `is_atomic` here?
In the `Iload` sequence below, we are outputting `dmb ishld` in some cases where `is_atomic` is set.

File: asmcomp/arm64/emit.mlp [hunk 2]
```diff
@@ -755,8 +752,6 @@ let emit_instr env i =
         emit_stack_adjustment (-n);
         env.stack_offset <- env.stack_offset + n
     | Lop(Iload { memory_chunk; addressing_mode; is_atomic }) ->
-        if is_atomic then ` dmb ishld\n`;
-        let load_type = if is_atomic then "ldar" else "ldr" in
         let dst = i.res.(0) in
```
Comment (ctk21) [inline asmcomp/arm64/emit.mlp:757]: The code (I think) is assuming that `memory_chunk = Word_int || memory_chunk = Word_val || is_atomic = false`. Presumably there is no way (currently) that we end up with an atomic load of a `Double`.

```suggestion
        assert(memory_chunk = Word_int || memory_chunk = Word_val || is_atomic = false);
        let dst = i.res.(0) in
```

File: asmcomp/arm64/emit.mlp [hunk 3]
```diff
@@ -767,34 +762,38 @@ let emit_instr env i =
...
+            end else
+              `	ldr	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+        | Double ->
+            `	ldr	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         end
-    | Lop(Istore(size, addr, initialization)) ->
-        (* memory model barrier for non-initializing store *)
-        if not initialization then ` dmb ishld\n`;
+    | Lop(Istore(size, addr, assignment)) ->
         let src = i.arg.(0) in
```
Comment (ctk21) [inline asmcomp/arm64/emit.mlp:794]: The code (I think) is assuming that `size = Word_int || size = Word_val || assignment = false`. 
```suggestion
        assert(size = Word_int || size = Word_val || assignment = false);
        let src = i.arg.(0) in
```

**Review by ctk21 — COMMENTED** [review]
Thanks, looks better than what was there. 

I think the lack of counting for the barrier that can be added in `Iload` is a bug. 

The asserts might be better as comments, not sure what's best.

File: asmcomp/arm64/emit.mlp (hunk 2 above)
Comment (xavierleroy) [reply asmcomp/arm64/emit.mlp:757]: OK for the assertion.  Atomic loads are generated only for the translation of `Atomic.get`, so only at `Word` type.

File: asmcomp/arm64/emit.mlp (hunk 3 above)
Comment (xavierleroy) [reply asmcomp/arm64/emit.mlp:794]: Noooo!  There are plenty of assignments at other sizes, e.g. in byte arrays, flat float arrays, bigarrays.  But we agreed (during the memory model meeting) that these assignments will not follow the Multicore OCaml memory model, so no barrier is needed.

Comment (ctk21) [reply asmcomp/arm64/emit.mlp:794]: I take it back, you are correct, bombing the compilation is offensive. 
Maybe just a comment documenting that those assignments will not follow the Multicore OCaml memory model is the thing to do? 
(I can just imagine discussions appearing on github if we don't document in the code :) 
```suggestion
        (* NB: assignments other than Word_int and Word_val do not follow the
        Multicore OCaml memory model and so do not emit a barrier *)
        let src = i.arg.(0) in
```

File: asmcomp/arm64/emit.mlp (hunk 1 above)
Comment (xavierleroy) [reply asmcomp/arm64/emit.mlp]: You're right.  Fixed in 8cdbf79 .

File: asmcomp/arm64/emit.mlp (hunk 3 above)
Comment (xavierleroy) [reply asmcomp/arm64/emit.mlp:794]: OK for the comment.  Memory model discussions tend to persist and repeat whatever we do :-),  but the comment is a start.

File: asmcomp/arm64/emit.mlp [hunk 4]
```diff
@@ -767,34 +764,40 @@ let emit_instr env i =
...
         | Single ->
-            `	{emit_string load_type}	s7, {emit_addressing addressing_mode base}\n`;
+            `	ldr	s7, {emit_addressing addressing_mode base}\n`;
             `	fcvt	{emit_reg dst}, s7\n`
-        | Word_int | Word_val | Double ->
-            `	{emit_string load_type}	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+        | Word_int | Word_val ->
+            if is_atomic then begin
+              assert (addressing_mode = Iindexed 0);
+              `	ldar	{emit_reg dst}, [{emit_reg i.arg.(0)}]\n`
```
Comment (ctk21) [inline asmcomp/arm64/emit.mlp:785]: So I was double checking the compilation of this:
```ocaml
let code0 x y = Atomic.set x 1 ; Atomic.get y
```
and I see:
```
_camlSBA__code0_787:
        sub     sp, sp, #0x10
        str     x30, [sp, #0x8]
        mov     x20, x1
        orr     x1, xzr, #0x3
        mov     x19, sp
        ldr     x16, [x28, #0x30]
        mov     sp, x16
        bl      _caml_atomic_exchange
        mov     sp, x19
        ldar    x0, [x20]
        ldr     x30, [sp, #0x8]
        add     sp, sp, #0x10
        ret
```
Shouldn't the atomic load give us:
```
        dmb     ishld
        ldar    x0, [x20]
```
as per #10995. (and I think clarified here: https://github.com/ocaml/ocaml/pull/10972#issuecomment-1032940299)

I think the change that is needed is:
```suggestion
              ` dmb ishld\n`
              `	ldar	{emit_reg dst}, [{emit_reg i.arg.(0)}]\n`
```

Comment (xavierleroy) [reply asmcomp/arm64/emit.mlp:785]: Yes, sure.  I thought I pushed that change already?  Apparently not :-(

Comment (xavierleroy) [reply asmcomp/arm64/emit.mlp:785]: PS.  When you're happy with this PR, I need to clean up the history!

**Comment by ctk21** [comment]
Great, I think we're there (although no doubt there might be more to do). 

Let's merge this and get the discussion back on ocaml#10972

File: asmcomp/arm64/emit.mlp (hunk 4 above)
Comment (ctk21) [reply asmcomp/arm64/emit.mlp:785]: ouch, apologies, merged it up. 
Is it more than squashing?

Comment (xavierleroy) [reply asmcomp/arm64/emit.mlp:785]: Just squashing.  We can do it when merging https://github.com/ocaml/ocaml/pull/10972

**Pull request merged**